runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -252,405 +252,256 @@ doubleMaxMinHelper(TR::Node *node, TR::CodeGenerator *cg, bool isMaxOp)
-      }
-
-   stringCountOffset  = cg->fej9()->getInstanceFieldOffsetIncludingHeader("Ljava/lang/String;", "count", "I" , comp->getCurrentMethod());
-   valueContentOffset = TR::Compiler->om.contiguousArrayHeaderSizeInBytes();
-
-   TR::LabelSymbol * processNext16Bytes = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
-   TR::LabelSymbol * doneLabel          = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
-   TR::LabelSymbol * handleInvalidChars = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
-   TR::LabelSymbol * coreLoopSetup      = TR::LabelSymbol::create(cg->trHeapMemory(),cg);
-
-   // Load this.value for source string and target string
-   if (usesCompressedrefs)
-      {
-      cursor =               generateRXInstruction          (cg, TR::InstOpCode::LLGF, node, rSrcVal, generateS390MemoryReference(rSrcBase, stringValOffset, cg));
-      if (shiftAmount != 0 ) generateRSInstruction          (cg, TR::InstOpCode::SLLG, node, rSrcVal, rSrcVal, shiftAmount);
-
-      cursor =               generateRXInstruction          (cg, TR::InstOpCode::LLGF, node, rTgtVal, generateS390MemoryReference(rTgtBase, stringValOffset, cg));
-      if (shiftAmount != 0 ) generateRSInstruction          (cg, TR::InstOpCode::SLLG, node, rTgtVal, rTgtVal, shiftAmount);
-      }
-   else
-      {
-      cursor =               generateRXInstruction          (cg, TR::InstOpCode::getLoadOpCode(), node, rSrcVal, generateS390MemoryReference(rSrcBase, stringValOffset, cg));
-      cursor =               generateRXInstruction          (cg, TR::InstOpCode::getLoadOpCode(), node, rTgtVal, generateS390MemoryReference(rTgtBase, stringValOffset, cg));
-      }                                                                                                                                                                                iComment("<- output.value");
-
-   // Load this.count. Skip 0 check since since that in done in Java
-   cursor =                  generateRXInstruction          (cg, is64 ? TR::InstOpCode::LLGF : TR::InstOpCode::L, node, rLen, generateS390MemoryReference(rSrcBase, stringCountOffset, cg));                     iComment("<- input.count");
-
-   // Bail out if string is larger than INT_MAX32/2 since # string to # byte conversion will cause overflow.
-   // We could have used 64 bit reg instructions and reg for 31bit JVM with proper sign/zero extension,
-   // but using this keeps the icache footprint down by a little bit. Also enables us to use the HPRs.
-   if (!is64)
+   if (is64)
       {
-      cursor =               generateRIInstruction          (cg, TR::InstOpCode::TMLH, node, rLen, (uint16_t) 0x8000);
-      cursor =               generateS390BranchInstruction  (cg, TR::InstOpCode::BRC , TR::InstOpCode::COND_MASK2, node, handleInvalidChars);
-      cursor->setStartInternalControlFlow();
+      generateRRInstruction(cg, TR::InstOpCode::LLGFR, node, lengthRegister, lengthRegister);
```

Comment:
Why exactly is this needed? Looking at the instruction selection from https://github.com/eclipse/openj9/pull/2098#issuecomment-395776065 I see that all uses of this register can be made in a 32-bit context. We seem to be forcing it to be 64-bit here however.

Reply:
cf269b9